File: fast_and_slow_pointer/happy-number.js
const find_happy_number = (num) => {
    let l = find_square_sum(num);
    let r = find_square_sum(l);

    while (l !== r){
        if (l === 1 || r === 1) return true
        l = find_square_sum(l);
        r = find_square_sum(find_square_sum(r))
    }

    return false;
}

const find_square_sum = (num) => {
    let sum = 0;
    while (num > 0) {
        let digit = num % 10;
        sum += digit ** 2;
        num = Math.floor(num / 10);
    }
    return sum;
}

console.log(`${find_happy_number(23)}`) // true
console.log(`${find_happy_number(12)}`) // false

File: fast_and_slow_pointer/python/happy-number.py
def find_happy_number(num):
    slow, fast = num, sum_of_squares(sum_of_squares(num))
    while slow != fast:
        if slow == 1 or fast == 1:
            return True
        slow = sum_of_squares(slow)
        fast = sum_of_squares(sum_of_squares(fast))
    return False


def sum_of_squares(num):
    _sum = 0
    for i in str(num):
        _sum += int(i) ** 2
    return _sum


if __name__ == '__main__':
    print(find_happy_number(23))
    print(find_happy_number(12))
